firedrake/preconditioners/asm.py

```diff
@@ -0,0 +1,98 @@
+import abc
+
+from pyop2.mpi import COMM_SELF
+from firedrake.preconditioners.base import PCBase
+from firedrake.petsc import PETSc
+from firedrake.dmhooks import get_function_space
+
+__all__ = ("ASMPatchPC", "ASMLinesmoothPC")
+
+
+class ASMPatchPC(PCBase):
+    ''' PC for PETSc PCASM
+
+    should implement:
+    - :meth:`get_patches`
+    '''
+
+    _prefix = "pc_asm_"
```

Comment:
Make this an `@abstractproperty` so that concrete classes have to implement it?

Reply:
apparently abstractproperty is deprecated, but I did it anyway